core/state/statedb.go

```diff
@@ -145,9 +145,28 @@ type StateDB struct {
 	// Testing hooks
 	onCommit func(states *triestate.Set) // Hook invoked when commit is performed
 
+	// arbitrum: cleanup hook
+	release  func()
+	released bool
+
 	deterministic bool
 }
 
+func (s *StateDB) SetRelease(release func()) {
```

Comment:
Let's talk about why finalizers don't work.
If not finalizers - I think I prefer your original solution returning finalizers when creating StateDB.
When upstream changes - it's best if we don't have to do anything (which was the idea of finalizers). Returning a new param will make us have to fix many merge conflicts which is a pain.
However, current approach could be worse - if some new place calls StateAtBlock we'll probably just not notice it, not add a defer state.Release where needed and end up with a memory leak.